discord/state.py

```diff
@@ -137,7 +139,200 @@ def done(self) -> None:
+
+    async def wait(self) -> List[Member]:
+        future = self.loop.create_future()
+        self.waiters.append(future)
+        try:
+            return await future
+        finally:
+            self.waiters.remove(future)
+
+    def get_future(self) -> asyncio.Future[List[Member]]:
+        future = self.loop.create_future()
+        self.waiters.append(future)
+        return future
+
+    def done(self) -> None:
+        for future in self.waiters:
+            if not future.done():
+                future.set_result(self.buffer)
+
+        try:
+            del self.state._scrape_requests[self.guild.id]
+        except KeyError:
+            pass
+
+    def start(self):
+        self.loop.create_task(self.wrapper())
+
+    async def wrapper(self):
+        try:
+            await self.scrape()
+        except RuntimeError:
+            _log.warning('Member list scraping failed for %s.', self.guild.id)
+            self.buffer = None
+        finally:
+            self.done()
+
+    async def scrape(self):
+        self.subscribing = True
+        delay = self.delay
+        channels = self.channels
+        guild = self.guild
+        ws = self.ws
+
+        while self.subscribing:
+            requests = {}
+            for channel in channels:
+                ranges = self.get_current_ranges()
+                if not ranges:
+                    self.subscribing = False
+                    break
+                requests[channel] = ranges
+
+            if not requests:
+                raise RuntimeError('Unable to get channels')
+
+            def predicate(data):
+                if int(data['guild_id']) != guild.id:
+                    return False
+
+                ranges = tuple(chain(requests.values()))
+                ranges_returned = tuple(chain(op.get('ranges', []) for op in data['ops']))
+                return any(range in ranges_returned for range in ranges)
+
+            _log.debug('Subscribing to %s ranges for guild %s.', requests, guild.id)
+            await ws.request_lazy_guild(guild.id, channels=requests)
+
+            try:
+                await asyncio.wait_for(ws.wait_for('GUILD_MEMBER_LIST_UPDATE', predicate), timeout=15)
+            except asyncio.TimeoutError:
+                _log.debug('Guild %s timed out waiting for a GUILD_MEMBER_LIST_UPDATE.', guild.id)
+                r = tuple(requests.values())[-1][-1]
+                if self.limit in range(r[0], r[1]) or self.limit < r[1]:
+                    self.subscribing = False
+                    break
+                else:
+                    raise RuntimeError('No response')
+
+            await asyncio.sleep(delay)
+
+            r = tuple(requests.values())[-1][-1]
+            if self.limit in range(r[0], r[1]) or self.limit < r[1]:
+                self.subscribing = False
+                break
+
+        if not self.chunk:  # Stop unnecessary GUILD_MEMBER_LIST_UPDATEs
+            await ws.request_lazy_guild(guild.id, channels={})
```

Comment:
On line 724, you use self.ws, what is `ws` and why isn't it `self.ws` here?

Reply:
If `self.ws` is accessed more than once in the same function, it's assigned to a local variable for performance reasons.  
https://stackoverflow.com/questions/28597014/python-why-is-accessing-instance-attribute-slower-than-local